lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -744,96 +738,122 @@ static bool typeContainsReset(FIRRTLType type) {
-            // - `make_early_inc_range` since `getUses()` is invalidated upon
-            //   `use.set(...)`.
-            // - `drop_begin` such that the first use can keep the original op.
-            auto newOp = builder.create<InvalidValueOp>(type);
-            use.set(newOp);
-          }
-        })
-
-        .Case<SubfieldOp>([&](auto op) {
-          // Associate the input bundle's resets with the output field's resets.
-          auto bundleType = op.getInput().getType();
-          auto index = op.getFieldIndex();
-          traceResets(op.getType(), op.getResult(), 0,
-                      bundleType.getElements()[index].type, op.getInput(),
-                      getFieldID(bundleType, index), op.getLoc());
-        })
 
-        .Case<SubindexOp, SubaccessOp>([&](auto op) {
-          // Associate the input vector's resets with the output field's resets.
-          //
-          // This collapses all elements in vectors into one shared element
-          // which will ensure that reset inference provides a uniform result
-          // for all elements.
-          //
-          // CAVEAT: This may infer reset networks that are too big, since
-          // unrelated resets in the same vector end up looking as if they were
-          // connected. However for the sake of type inference, this is
-          // indistinguishable from them having to share the same type (namely
-          // the vector element type).
-          auto vectorType = op.getInput().getType();
-          traceResets(op.getType(), op.getResult(), 0,
-                      vectorType.getElementType(), op.getInput(),
-                      getFieldID(vectorType), op.getLoc());
-        })
+  SmallVector<std::pair<FModuleOp, SmallVector<Operation *>>> moduleToOps;
 
-        .Case<RefSubOp>([&](RefSubOp op) {
-          // Trace through ref.sub.
-          auto aggType = op.getInput().getType().getType();
-          uint64_t fieldID =
-              TypeSwitch<FIRRTLBaseType, uint64_t>(aggType)
-                  .Case<FVectorType>([](auto type) { return getFieldID(type); })
-                  .Case<BundleType>([&](auto type) {
-                    return getFieldID(type, op.getIndex());
-                  });
-          traceResets(op.getType(), op.getResult(), 0, op.getResult().getType(),
-                      op.getInput(), fieldID, op.getLoc());
-        });
+  for (auto module : circuit.getOps<FModuleOp>())
+    moduleToOps.push_back({module, {}});
+
+  mlir::parallelForEach(circuit.getContext(), moduleToOps, [](auto &e) {
+    e.first.walk([&](Operation *op) {
+      // We are only interested in operations which are related to abstract
+      // reset.
+      if (llvm::any_of(
+              op->getResultTypes(),
+              [](mlir::Type type) { return typeContainsReset(type); }) ||
+          llvm::any_of(op->getOperandTypes(), typeContainsReset))
+        e.second.push_back(op);
```

Comment:
Is it thread safe to update the vector `moduleToOps` in parallel ? If it is, then this is great.